flink-runtime/src/main/java/org/apache/flink/runtime/entrypoint/ClusterEntrypoint.java

```diff
@@ -579,6 +580,19 @@ protected static Configuration loadConfiguration(
 
     public static void runClusterEntrypoint(ClusterEntrypoint clusterEntrypoint) {
 
+        // Register the signal handler so that we could have the chance to execute the shutdown
+        // supplier before handling the signal.
+        SignalHandler.register(
+                LOG,
+                () -> {
+                    if (clusterEntrypoint.isShutDown.get()) {
+                        LOG.info("Waiting for concurrent shutDownAsync finished.");
+                        return clusterEntrypoint.terminationFuture.thenAccept(ignored -> {});
+                    } else {
+                        return FutureUtils.completedVoidFuture();
```

Comment:
Maybe we could have a more graceful shutdown here in the future. For example, `stopClusterServices` without cleaning HA data, `cleanupDirectories`. In this PR, we keep the same behavior with before. When received `SIGTERM`, the JVM will directly exit. It is enough since the `cleanupDirectories` will be done in the `shutdownHook`.